chain/chunks/src/lib.rs

```diff
@@ -491,14 +495,68 @@ impl ShardsManager {
         Ok(())
     }
 
+    pub fn store_partial_encoded_chunk(
+        &mut self,
+        known_header: &BlockHeader,
+        partial_encoded_chunk: PartialEncodedChunk,
+    ) {
+        // Remove old partial_encoded_chunks
+        let encoded_chunks = &self.encoded_chunks;
+        self.stored_partial_encoded_chunks
+            .retain(|&height, _| encoded_chunks.height_within_front_horizon(height));
+
+        let header = partial_encoded_chunk
+            .clone()
+            .header
+            .expect("PartialEncodedChunkMessage always have header");
+        let height = header.inner.height_created;
+        let shard_id = header.inner.shard_id;
+        if self.encoded_chunks.height_within_front_horizon(height) {
+            let runtime_adapter = &self.runtime_adapter;
+            let heights =
+                self.stored_partial_encoded_chunks.entry(height).or_insert(HashMap::new());
+            heights
+                .entry(shard_id)
+                .and_modify(|stored_chunk| {
+                    let epoch_id = runtime_adapter
```

Comment:
Also add some basic sanity test for this behavior.

Reply:
++